privacyidea/lib/tokens/tantoken.py

```diff
@@ -0,0 +1,218 @@
+from privacyidea.lib.crypto import geturandom, hash
+import binascii
+
+log = logging.getLogger(__name__)
+DEFAULT_COUNT = 100
+SALT_LENGTH = 4
+
+
+class TANACTION(object):
+    TANTOKEN_COUNT = "tantoken_count"
+
+
+class TanTokenClass(PaperTokenClass):
+    """
+    The TAN token allows to print out the next e.g. 100 OTP values.
+    This sheet of paper can be used to authenticate and strike out the used
+    OTP values. In contrast to the paper token, the OTP values of the TAN token
+    can be used randomly.
+    """
+
+    @log_with(log)
+    def __init__(self, db_token):
+        """
+        This creates a new TAN token object from a DB token object.
+
+        :param db_token: instance of the orm db object
+        :type db_token:  orm object
+        """
+        TokenClass.__init__(self, db_token)
+        self.set_type(u"tan")
+        self.hKeyRequired = False
+
+    @staticmethod
+    def get_class_type():
+        """
+        return the token type shortname
+
+        :return: 'paper'
+        :rtype: string
+        """
+        return "tan"
+
+    @staticmethod
+    def get_class_prefix():
+        """
+        Return the prefix, that is used as a prefix for the serial numbers.
+        :return: PITN
+        """
+        return "PITN"
+
+    @staticmethod
+    @log_with(log)
+    def get_class_info(key=None, ret='all'):
+        """
+        returns a subtree of the token definition
+
+        :param key: subsection identifier
+        :type key: string
+        :param ret: default return value, if nothing is found
+        :type ret: user defined
+        :return: subsection if key exists or user defined
+        :rtype: dict or scalar
+        """
+        res = {'type': 'tan',
+               'title': 'TAN Token',
+               'description': 'TAN: TANs printed on a sheet '
+                              'of paper.',
+               'init': {},
+               'config': {},
+               'user':  ['enroll'],
+               # This tokentype is enrollable in the UI for...
+               'ui_enroll': ["admin", "user"],
+               'policy': {
+                   SCOPE.ENROLL: {
+                       TANACTION.TANTOKEN_COUNT: {
+                           "type": "int",
+                           "desc": _("The number of OTP values, which are "
+                                     "printed on the paper.")
+                       }
+                   }
+               }
+               }
+
+        if key:
+            ret = res.get(key, {})
+        else:
+            if ret == 'all':
+                ret = res
+        return ret
+
+    def update(self, param, reset_failcount=True):
+        tan_dict = {}
+        if "tans" in param:
+            # init tokens with tans
+            tans = param.get("tans").split()
+            tan_dict = {k: v for k, v in enumerate(tans)}
+            # Avoid to generate TANs in the superclass PaperToken, since we get the from params
+            param["papertoken_count"] = 0
```

Comment:
This does not seem to work, because PaperTokenClass uses
```
papertoken_count = int(param.get("papertoken_count") or DEFAULT_COUNT)
```
and since ``param.get("papertoken_count")`` evaluates to 0, DEFAULT_COUNT is used.

... if we want to be especially hacky, we could use ``param["papertoken_count"] = "0"`` ... :)

Reply:
Well I think actually in the line

~~~~python
    papertoken_count = int(param.get("papertoken_count") or DEFAULT_COUNT
~~~~

what we mean is:

~~~~python
    if "papertoken_count" not in param:
       papertoken_count = DEFAULT_COUNT
    else:
       papertoken_count = param.get("papertoken_count")
~~~~

Or

~~~~python
    papertoken_count = DEFAULT_COUNT
    if "papertoken_count" in param:
        papertoken_count = param.get("papertoken_count")
~~~~

But since usually param really contains strings, since they are passed from the API, we could also use the string "0".

Whhat do you think?

Reply:
Hm, I think both alternatives you mentioned are equivalent to
```python
papertoken_count = param.get("papertoken_count", DEFAULT_COUNT)
```
(well, we need to use ``int(...)`` as well though)

The difference to the current behavior, however, is that ``papertoken_count`` can be the empty string currently (we would choose DEFAULT_COUNT then). With the changed line, this is not allowed anymore. Do you think we should allow the empty string?

Reply:
I think you are right. I do not know if there was a scenario, where we explicitly wanted to have ``DEFAULT_COUNT`` paper tokens, although the user specified ``papertoken_count = 0``

If will got with

    int(param.get("papertoken_count", DEFAULT_COUNT))